Adds webpack and per-page javascript

anselmbradford opened the pull request:
Adds webpack and per-page javascript.

Examples of module scope:
- Add a global-level module in `/src/static/js/routes/common.js`
- Add a template-level module in (e.g.) `/src/static/js/routes/offices/single.js`
- Add a page-level module in (e.g.) `/src/static/js/routes/contact-us/index.js`

## Additions

- Added webpack module loader for per-page JavaScript.
- Added script manifest for passing page-specific script paths to runtime page-script loader.
- Added external-site page-specific script.

## Removals

- Removed unused exportsOverride section, which was an artifact of the grunt bower task.
- Removed browserify, watchify, and browserify-shim dependencies.

## Changes

- Moved handlebars from npm to bower.
- Added jQuery CDN with fallback to head to satisfy GTM requirements. Closes https://github.com/cfpb/cfgov-refresh/issues/587

## Testing

- `gulp build` should pass. Add a console.log statement to js/routes/common.js, js/routes/contact-us, and js/routes/offices and run `gulp build` and check console in `/contact-us/` and `/offices/`.

## Review

- @jimmynotjim 
- @sebworks 
- @KimberlyMunoz 

## Todos

- I had to deliver all JS without a `.min` filename, even though it's minified. Ideally, we'll need a dev vs production configuration that handles appending the filenames, but also allows uncompressed delivery.
- I had to remove the .bowererr lookup for the vendor directory and hardcode it instead, as the fs module was somehow making it into common.js and breaking it when that was there.
- The routes-manifest is so we don't try to dynamically load a page-specific script that isn't there. Ideally, we could have the backend check if a script exists and build the page-javascript info in the base layout, instead of doing it via node. But I wasn't able to figure out how to do this with our available jinja syntax.

KimberlyMunoz commented:
Should we merge #965 before we merge this one?

sebworks commented:
Should we be using dashes instead of dot notation for the config file name?

anselmbradford commented:
> Should we merge #965 before we merge this one?

Yea probably.

> Should we be using dashes instead of dot notation for the config file name?

So that's just based on how it was shown in the webpack docs, but it's by no means [consistent](https://github.com/webpack/react-starter/blob/master/webpack-production.config.js). I'd be fine with updating it to dashes. We have three different JS filename styles: dashes, dots, and camelCase. Removing one seems good. Fixed in https://github.com/cfpb/cfgov-refresh/commit/4ebf17b6cfe71688d658c24f9822380bb9fff858

sebworks commented on gulp/tasks/scripts.js:
```diff
@@ -1,83 +1,41 @@
+    } ) );
 }
 
 gulp.task( 'scripts', function() {
```
Should this have a JSDoc comment?

sebworks commented on gulp/config.js:
```diff
@@ -1,16 +1,7 @@
-  lib:  JSON.parse( fs.readFileSync( './.bowerrc' ) ).directory, // eslint-disable-line no-sync, no-inline-comments, max-len
-  test: './test'
-};
+var paths = require( '../config/environment-production' ).paths;
```
Shouldn't we have a config file for the `dev` environment?

anselmbradford replied:
Yeah, this is just named `-production` in anticipation of having a `-development` version.

sebworks replied:
I think a TODO might be in order.

anselmbradford replied:
Thinking about it further, environment settings we currently have would be the same on production vs dev, but the webpack config would be different. I think I'll just remove `-production` for now.

anselmbradford replied:
Fixed in https://github.com/cfpb/cfgov-refresh/commit/80c7df834815458d8c00da731c7f803de63ced6e

sebworks commented on src/static/js/routes/common.js:
```diff
@@ -4,40 +4,31 @@
+require( '../modules/nav-primary.js' ).init();
+require( '../modules/secondary-nav-toggle' ).init();
+require( '../modules/mobile-primary-nav.js' ).init();
+require( '../modules/desktop-primary-nav.js' ).init();
```
`desktop-primary-nav.js` and `mobile-primary-nav.js` no longer exist.

anselmbradford replied:
Interesting. I wonder why it didn't complain about that.

anselmbradford replied:
Fixed in https://github.com/cfpb/cfgov-refresh/commit/322a52d2643c038cb1d8b0a5ecee4450ca9ea8da

sebworks commented on src/static/js/modules/per-page-js-loader.js:
```diff
@@ -0,0 +1,34 @@
+ * if the page has a page-specific script available.
+ */
+function init() {
+  var jsLoader = require( '../modules/util/js-loader' );
```
Why define `jsManifest` outside the function and `jsLoader` inside the function?

anselmbradford replied:
Fixed in https://github.com/cfpb/cfgov-refresh/commit/a28eaa72b807d831b51e1db642eba0678fc42aac

sebworks commented on src/static/js/modules/per-page-js-loader.js:
```diff
@@ -0,0 +1,34 @@
+ */
+function init() {
+  var jsLoader = require( '../modules/util/js-loader' );
+  var rawJson = document.getElementById( 'page-javascript' ).innerHTML;
```
Is there a better way to handle parsing and loading scripts than `innerHTML`/JSON.Parse?

sebworks replied:
Why not handle creating the per page js bundles in advance on the server?

anselmbradford replied:
So the bundles themselves are created in advance, but do you mean the lookup for whether they exist or not? Because, yes, that should happen on the server and I have a TODO [here](https://github.com/cfpb/cfgov-refresh/pull/968/files#diff-a2823c2782d047ce8761801fa918298dR198) for that. I couldn't figure out how to do it with our existing available jinja commands. Ideally it would do something like inject the per-page JS URL to load, and then the per-page-js-loader would load that URL if it was present on the page and do nothing if it wasn't. Let me know if you have any ideas!

jimmynotjim commented on package.json:
```diff
@@ -5,12 +5,11 @@
+    "dateformat": "^1.0.11",
     "del": "^1.2.0",
-    "fs": "0.0.2",
+    "fs": "latest",
```
Do we really want latest?

anselmbradford replied:
I realized we don't need `fs`, it's built into node. I mean what is [this](https://www.npmjs.com/package/fs)???

Fixed in https://github.com/cfpb/cfgov-refresh/commit/f116853001dad01f81ff76c22fe0129f31badb49

jimmynotjim commented on src/_layouts/base.html:
```diff
@@ -99,6 +99,14 @@
+    {# TODO: jQuery CDN is provided in the head to satisfy
+             Google Tag Manager (GTM) requirements.
+             Ideally GTM would handle its own dependency management
+             and initializing jQuery could be moved to the footer. #}
```
Wouldn't that lead to jQuery being loaded twice?

anselmbradford replied:
No, if jQuery checks if it's needed before loading it wouldn't.

jimmynotjim commented on src/static/js/routes/careers/working-at-cfpb/index.js:
```diff
@@ -6,16 +6,6 @@
-
-module.exports = { init: init };
+// MobileCarousel with 599px breakpoint.
+var mobileCarousel = new MobileCarousel( 599 );
```
Should the breakpoint come from the breakpoint config file?

anselmbradford replied:
Fixed in https://github.com/cfpb/cfgov-refresh/commit/8b200a8afd0c4c600fcda4088fc777f788e54617

sebworks replied:
You shouldn't have to pass any value. It should default to [599](https://github.com/cfpb/cfgov-refresh/blob/flapjack/src/static/js/modules/classes/MobileCarousel.js)?

anselmbradford commented:
Talked with @sebworks about the dynamic loading of the page scripts and the un-ideal world of passing json to the page and parsing it. Decided to strip out the routes manifest stuff and inject the template and page-level scripts directly into the page. This simplifies things and works nicely I think. Those scripts get combined with the page request. They will also be cached by the `include`, if this proves problematic we can add `with context` to the includes in `base.html`. If we discover a particular page has a large script that holds up the UI thread on page load, that script can dynamically load its contents from within the page-script... giving the best of both worlds. Fixes added in https://github.com/cfpb/cfgov-refresh/commit/b451ddb14bcb0bcbebc0fd8e814058007f784f6d Please review and sign off. Thanks!

jimmynotjim commented:
Why not switch to async loading for non-critical scripts like those for the nav and eventually search bar?

anselmbradford commented:
> Why not switch to async loading for non-critical scripts like those for the nav and eventually search bar?

Sure, out of scope with this PR I'd say, but any of the page-specific or global scripts can just be shells for async loading their main contents.

jimmynotjim commented:
Sorry, didn't mean it was in the scope, just in reply to

>  If we discover a particular page has a large script that holds up the UI thread on page load, that script can dynamically load its contents from within the page-script

I should have put that quote in my original reply.

anselmbradford commented:
Added fix for `/contact-us/` page. When using `include` the included file is run through jinja. So this will cause problems anywhere with `{{` or other notation recognized by jinja. Therefore imported JS should be enclosed in `<% raw %> … <% endraw %>`. This can't be added outside the `include` or the `include` will not execute. SO... it needs to be added to the JS files themselves. I added the [banner-footer-webpack-plugin](https://github.com/dtothefp/banner-footer-webpack-plugin), which allows injecting arbitrary text at the top and bottom of a file. This is used to add the raw tags to the JS after minification. Changes in https://github.com/cfpb/cfgov-refresh/commit/04065286a8c0117329f0e33961159d0dd0b86ec1

sebworks commented:
:+1:

jimmynotjim commented:
All good here :+1:

anselmbradford commented:
Rebased, but retained commit to remove route manifest. If that's ever needed for reference.
